eth/tracers/api_blockResult.go

```diff
@@ -0,0 +1,347 @@
+}
+
+type traceEnv struct {
+	config *TraceConfig
+
+	// rMu lock is used to protect txs executed in parallel.
+	rMu      sync.Mutex
+	signer   types.Signer
+	state    *state.StateDB
+	tracer   *vm.StructLogger
+	blockCtx vm.BlockContext
+
+	// pMu lock is used to protect Proofs' read and write mutual exclusion,
+	// since txs are executed in parallel, so this lock is required.
+	pMu sync.Mutex
+	// sMu is required because of txs are executed in parallel,
+	// this lock is used to protect StorageTrace's read and write mutual exclusion.
+	sMu sync.Mutex
+	*types.StorageTrace
+	executionResults []*types.ExecutionResult
+}
+
+// GetBlockResultByNumberOrHash replays the block and returns the structured BlockResult by hash or number.
+func (api *API) GetBlockResultByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash, config *TraceConfig) (trace *types.BlockResult, err error) {
+	var block *types.Block
+	if number, ok := blockNrOrHash.Number(); ok {
+		block, err = api.blockByNumber(ctx, number)
+	}
+	if hash, ok := blockNrOrHash.Hash(); ok {
+		block, err = api.blockByHash(ctx, hash)
+	}
+	if err != nil {
+		return nil, err
+	}
+	if block.NumberU64() == 0 {
+		return nil, errors.New("genesis is not traceable")
+	}
+	if config == nil {
+		config = &TraceConfig{
+			LogConfig: &vm.LogConfig{
+				EnableMemory:     false,
+				EnableReturnData: true,
+			},
+		}
+	} else if config.Tracer != nil {
+		config.Tracer = nil
+		log.Warn("Tracer params is unsupported")
+	}
+
+	// create current execution environment.
+	env, err := api.createTraceEnv(ctx, config, block)
+	if err != nil {
+		return nil, err
+	}
+
+	return api.getBlockResult(block, env)
+}
+
+// Make trace environment for current block.
+func (api *API) createTraceEnv(ctx context.Context, config *TraceConfig, block *types.Block) (*traceEnv, error) {
+	parent, err := api.blockByNumberAndHash(ctx, rpc.BlockNumber(block.NumberU64()-1), block.ParentHash())
+	if err != nil {
+		return nil, err
+	}
+	reexec := defaultTraceReexec
+	if config != nil && config.Reexec != nil {
+		reexec = *config.Reexec
+	}
+	statedb, err := api.backend.StateAtBlock(ctx, parent, reexec, nil, true, true)
+	if err != nil {
+		return nil, err
+	}
+
+	env := &traceEnv{
+		config:   config,
+		signer:   types.MakeSigner(api.backend.ChainConfig(), block.Number()),
+		state:    statedb,
+		tracer:   vm.NewStructLogger(config.LogConfig),
+		blockCtx: core.NewEVMBlockContext(block.Header(), api.chainContext(ctx), nil),
+		StorageTrace: &types.StorageTrace{
+			RootBefore:    parent.Root(),
+			RootAfter:     block.Root(),
+			Proofs:        make(map[string][]hexutil.Bytes),
+			StorageProofs: make(map[string]map[string][]hexutil.Bytes),
+		},
+		executionResults: make([]*types.ExecutionResult, block.Transactions().Len()),
+	}
+
+	for _, coinbase := range []common.Address{block.Coinbase(), api.backend.Coinbase()} {
```

Comment:
why do we need `api.backend.Coinbase()`?

Reply:
BlockResult should have coinbase proof.

Reply:
This is great! We should use this for the coinbase address

Reply:
This hasn't been fixed

Reply:
I just tested this, `api.backend.Coinbase()` will return the correct coinbase address for Clique.

Reply:
I just tested this, `api.backend.Coinbase()` will return the correct coinbase address for Clique, but this approach will only work for locally produced blocks, not for blocks imported from other peers.

Reply:
changed.

Reply:
Use `api.backend.Engine().Author(block.Header())` is the right way.